utils/zerovec/src/ule/unvalidated.rs

```diff
@@ -0,0 +1,207 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::VarULE;
+use crate::{map::ZeroMapKV, VarZeroSlice, VarZeroVec, ZeroVecError};
+use alloc::boxed::Box;
+use core::fmt;
+use core::ops::Deref;
+
+/// A byte slice that is expected to be a UTF-8 string but does not enforce that invariant.
+///
+/// Use this type instead of `str` if you don't need to enforce UTF-8 during deserialization. For
+/// example, strings that are keys of a map don't need to ever be reified as `str`s.
+///
+/// The main advantage of this type over `[u8]` is that it serializes as a string in
+/// human-readable formats like JSON.
+///
+/// # Examples
+///
+/// Using an [`UnvalidatedStr`] as the key of a [`ZeroMap`]:
+///
+/// ```
+/// use zerovec::ZeroMap;
+/// use zerovec::ule::UnvalidatedStr;
+///
+/// let map: ZeroMap<UnvalidatedStr, usize> = [
+///     (UnvalidatedStr::from_str("abc"), 11),
+///     (UnvalidatedStr::from_str("def"), 22),
+///     (UnvalidatedStr::from_str("ghi"), 33)
+/// ].into_iter().collect();
+///
+/// let key = "abc";
+/// let value = map.get_copied_by(|uvstr| (**uvstr).cmp(key.as_bytes()));
```

Comment:
Optional: actually I just realised that you're comparing `UnvalidatedStr` and `str`, which both have an `as_bytes()`, so I think it would be best to do `uvstr.as_bytes().cmp(key.as_bytes())`